src/java_bytecode/generate_java_generic_method.cpp

```diff
@@ -0,0 +1,296 @@
+  // defined the symbol.
+  const std::string& signature_decoration=instantiation_decoration
+    (concrete_types);
+
+  // Check the method has not already been specialized
+  symbol_tablet::symbolst::const_iterator already_specialized=
+    symbol_table.symbols.find(
+      id2string(generic_method.name)+signature_decoration);
+  if(already_specialized!=symbol_table.symbols.end())
+    return already_specialized->second;
+
+  // Convert concrete types into a map for faster lookups
+  type_variable_instantiation_mapt concrete_type_map(concrete_types);
+
+  // Copy the generic method as a starting point for the specialization
+  symbolt specialized_method=generic_method;
+  code_typet &specialized_code=to_code_type(specialized_method.type);
+
+  // Handle a generic return type
+  typet &method_return_type=specialized_code.return_type();
+  if(is_java_generic_parameter(method_return_type))
+  {
+    instantiate_java_generic_parameter(
+      to_java_generic_parameter(method_return_type),
+      concrete_type_map);
+  }
+  else if(is_java_generic_type(method_return_type))
+  {
+    instantiate_java_generic_type(
+      to_java_generic_type(method_return_type),
+      concrete_type_map);
+  }
+
+  // Set the name of the specialized method, but ensure we remember the
+  // original method name because we need it for cleaning up internal symbol
+  // names...
+  const irep_idt original_name=specialized_method.name;
+  specialized_method.name=
+    id2string(specialized_method.name)+signature_decoration;
+
+  for(code_typet::parametert &parameter : specialized_code.parameters())
+  {
+    // Update the symbol name of the parameter to match the newly decorated
+    // name of the specialized method.
+    decorate_identifier(parameter, ID_C_identifier, original_name,
+      signature_decoration);
+
+    typet &parameter_type=parameter.type();
+
+    if(is_java_generic_parameter(parameter_type))
+    {
+      instantiate_java_generic_parameter(
+        to_java_generic_parameter(parameter_type),
+        concrete_type_map);
+    }
+    else if(is_java_generic_type(parameter_type))
+    {
+      instantiate_java_generic_type(
+        to_java_generic_type(parameter_type),
+        concrete_type_map);
+    }
+  }
+
+  // Updated the body of the specialized method so that all references to
+  // symbols in the method body refer to the specialized method and not the
+  // generic method.
+  // FIXME: Probably also need to update references to fields to point to the
+  // FIXME: specialized class fields, rather than the generic class fields?
+  auto body_iterator=specialized_method.value.depth_begin();
+  const auto body_end=specialized_method.value.depth_end();
+
+  while(body_iterator!=body_end)
+  {
+    if(body_iterator->id()==ID_symbol)
+    {
+      exprt &symbol_expr=body_iterator.mutate();
+      typet &symbol_type=symbol_expr.type();
+      if(is_java_generic_parameter(symbol_type))
```

Comment:
Nit: Consider pulling out this if and the following else if to avoid duplication:

```C++
void instantiate_generic_types(const typet &parameter_type, const concrete_type_map)
{
  if(is_java_generic_parameter(parameter_type))
  {
     instantiate_java_generic_parameter(
       to_java_generic_parameter(parameter_type),
       concrete_type_map);
   }
   else if(is_java_generic_type(parameter_type))
   {
     instantiate_java_generic_type(
       to_java_generic_type(parameter_type),
       concrete_type_map);
    }
}
```

Reply:
Ah yes, good point - should have seen that myself. Will do.